src/lib/signature.ts

```diff
@@ -208,10 +225,51 @@ class Signature extends CircuitValue {
    */
   verify(publicKey: PublicKey, msg: Field[]): Bool {
     const point = publicKey.toGroup();
-    let e = Scalar.fromBits(
-      Poseidon.hash(msg.concat([point.x, point.y, this.r])).toBits()
+    let h = hashWithPrefix(
+      prefixes.signatureTestnet,
+      msg.concat([point.x, point.y, this.r])
     );
-    let r = point.scale(e).neg().add(Group.generator.scale(this.s));
+    // TODO: Scalar.fromBits interprets the input as a "shifted scalar"
+    // therefore we have to use scaleShifted which is very inefficient
+    let e = Scalar.fromBits(h.toBits());
+    let r = scaleShifted(point, e).neg().add(Group.generator.scale(this.s));
     return Bool.and(r.x.equals(this.r), r.y.toBits()[0].equals(false));
   }
+
+  /**
+   * Decodes a base58 encoded signature into a {@link Signature}.
+   */
+  static fromBase58(signatureBase58: string) {
+    let { r, s } = SignatureBigint.fromBase58(signatureBase58);
+    return Signature.fromObject({
+      r: Field(r),
+      s: Scalar.fromJSON(s.toString()),
+    });
+  }
+  /**
+   * Encodes a {@link Signature} in base58 format.
+   */
+  toBase58() {
+    let r = this.r.toBigInt();
+    let s = BigInt(this.s.toJSON());
+    return SignatureBigint.toBase58({ r, s });
+  }
+}
+
+// performs scalar multiplication s*G assuming that instead of s, we got s' = 2s + 1 + 2^255
+// cost: 2x scale by constant, 1x scale by variable
+function scaleShifted(point: Group, shiftedScalar: Scalar) {
+  let oneHalfGroup = point.scale(Scalar.fromBigInt(oneHalf));
+  let shiftGroup = oneHalfGroup.scale(Scalar.fromBigInt(shift));
+  return oneHalfGroup.scale(shiftedScalar).sub(shiftGroup);
```

Comment:
this is how we undo the scalar shift inside the SNARK

Reply:
why not just do

```
point.scale((shiftedScalar.sub(shift).mul(inv_2))
```

Reply:
oh I think I see, the scalars are in the wrong field :D

Reply:
yep exactly :D we don't have the foreign field arithmetic to do `sub` and `mul` on the scalar, inside the snark